### PR TITLE
Improve parsenum tests

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

**Comment by gperciva** [comment]
This ensures that we reach the appropriate `ASSERT_FAIL()` if we attempt to misuse the `PARSENUM` API.

**Comment by gperciva** [comment]
Some gcc thing, I'm investigating.

File: tests/parsenum/main.c [hunk 1]
```diff
@@ -151,7 +151,8 @@ test_assert_failure(const char * argv_1)
 	double f;
 
 	/* Which test should we attempt? */
-	PARSENUM(&assert_num, argv_1);
+	if (PARSENUM(&assert_num, argv_1))
+		exit(1);
```
[the remark this answers is not in the document]
Comment (cperciva) [reply tests/parsenum/main.c]: I think `exit(1)` is fine.  You're checking to make sure the right error message gets printed, so I think it should be fine...?

**Comment by gperciva** [comment]
Ready for review.

File: tests/parsenum/test_parsenum_improper_usage.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh -e
+
+# Check assertion failures, separating lines so that we automatically check
+# that test_parsenum is giving a non-zero exit code.
+output=$(! ./test_parsenum 1 2>&1)
+echo "$output" | grep -q "PARSENUM applied to signed integer without specified bounds"
```
Comment (cperciva) [inline tests/parsenum/test_parsenum_improper_usage.sh]: Is there a reason you don't want to use a temporary file here?  I'd say that
```
! ./test_parsenum 1 2>tmpfile
grep -q "..." tmpfile
```
is much cleaner than using `$()` and `echo`.

**Comment by gperciva** [comment]
Rebased to use a temporary file for the test script.

File: tests/parsenum/main.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (cperciva) [reply tests/parsenum/main.c]: Hmm, on second glance: While the `exit(1)` is right, wouldn't it be nice to at least print a warning about the bogus string we were passed?  e.g., `warnp("Parameter should be an error case between 1 and 4: %s", argv_1);` ?

Of course, that would mean adding `warnp` back to the test, but I don't think there's any critical need to keep it out...?

**Comment by gperciva** [comment]
Rebased to not take out `warnp`, and use that for the `argv_1` check.

**Pull request merged**